### ISSUE TITLE
Error with generic parameter with default value

### ISSUE THREAD

**Issue by Timmmm** [microsoft/pyright#3922]
Consider this code:

```python
class Foo:
    def __init__(self, a: int):
        pass

class FooDef(Foo):
    pass

F = TypeVar("F", bound=Foo)

def make(t: Type[F] = FooDef) -> F:
    return t(5)
```

This gives a type error for `t(5)` (using Pylance v2022.9.10):

![image](https://user-images.githubusercontent.com/376842/189172761-b7e7eec6-d629-454e-b685-8291eaf8e082.png)

This doesn't seem quite right. The parameter type also doesn't quite match the explicit annotation which is kind of weird:

![image](https://user-images.githubusercontent.com/376842/189172993-a8ca41fd-09ce-46cf-b368-862c26efe52a.png)

Would you expect this to work?

**Comment by Timmmm** [comment]
Even simpler:

```python
T = TypeVar("T", bound=str)
def make(t: Type[T] = str) -> T:
    return t()
```

**Comment by erictraut** [comment]
Parameters with generic types _and_ default arguments are problematic for the Python type system. Other Python type checkers (including mypy) don't allow this combination at all. For example, mypy reports the error:

```
error: Incompatible default for argument "t" (default has type "Type[FooDef]", argument has type "Type[F]")
```

Pyright allows this, but to make it type safe, the type of the parameter is implicitly a union between the TypeVar (`Type[F]`) and the type of the default argument (`Type[FooDef]`). This approach has caused enough confusion that I now regret implementing it. I've wondered whether it would be better to simply fall back to the approach that mypy has taken and make this combination illegal. This is an area where we should probably go back to the typing community and try to agree upon a standard behavior.

For now however, this is working as designed in pyright, so I don't consider it a bug.

**Comment by Timmmm** [comment]
Ah thanks for the explanation. I guess I can just cast it for now.
